### PR TITLE
OAuth Fixes

### PR DESCRIPTION
**Pull request by steve-gombos**
Fixes: #146  #144

Added the auth.py class to split out the ring oauth.  This is a rough pass at this, and probably need to work more with someone from HA to make sure this is in the right direction.  

Updated the Ring class _get_auth_token to use the new Auth class for the intial fetch, then refresh after that.

The only thing I don't understand is how you are managing the local pickle file and the session post to the Ring API.  If you could give me some clarity on this that would help, but basically right now I am just sticking the whole oauth token response into the pickle.  I think the end goal is to eventually have HA maintain this and just pass in the token to the Ring class.  Also I commented line 61 in __init__ because this was always wiping out the pickle, not sure if that was intended?

### CODE REVIEW

**Comment by tchellomello** [comment]
Wow @steve-gombos  You moved fast!!! Awesome, I'll take a look at the code tonight and let's see if we can ship it ASAP into HASS. Have you tested this code on your personal HASS instance for a couple of hours to see if that is stable? 

You rock and thanks so much the effort you put into this!

**Comment by Tosko4** [comment]
Thank you @steve-gombos and @tchellomello I bought a ring, and was wondering why it does not work.. was starting to doubt myself. Once reviewed, does this mean there will be a 0.103.5 release?

**Comment by tchellomello** [comment]
> Thank you @steve-gombos and @tchellomello I bought a ring, and was wondering why it does not work.. I was starting to doubt myself. Once reviewed, does this mean there will be a 0.103.5 release?

@Tosko4 Thanks for your kind words. We can't guarantee which version will be present, but we can work with @balloob to have it as soon as possible. In the worst-case scenario, you can apply a hotfix to address temporarily the issue.

**Comment by steve-gombos** [comment]
@tchellomello I have this enabled on my HA instance.  Will run for a bit and see how it goes.

**Comment by Tosko4** [comment]
Am I correct that this will also fix the 2FA authentication?

**Comment by steve-gombos** [comment]
@Tosko4 I did some rough testing with this on my account, but I did add a callback func param to the Ring class.  Since I am not super familiar with how to integrate this with HA, I haven't done anything on that front, though the idea is that you would get a prompt similar to something like the alexa_media component, input the code, it would then be able to pass that along to the Ring auth.

File: ring_doorbell/__init__.py [hunk 1]
```diff
@@ -29,7 +31,9 @@
 class Ring(object):
     """A Python Abstraction object to Ring Door Bell."""
 
-    def __init__(self, username, password, debug=False, persist_token=False,
+    def __init__(self, username, password,
```
Comment (balloob) [inline ring_doorbell/__init__.py]: I would suggest we change the Ring object to just pass in the auth as an object.

File: ring_doorbell/__init__.py [hunk 2]
```diff
@@ -90,18 +100,20 @@ def _process_cached_session(self):
     def _get_oauth_token(self):
```
Comment (balloob) [inline ring_doorbell/__init__.py]: In a future PR we should consider moving this logic out of this function as well. Separate the API logic from the Auth logic.

File: ring_doorbell/auth.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        token = self._oauth.refresh_token(OAuth.ENDPOINT)
+
+        if self.token_updater is not None:
+            self.token_updater(token)
+
+        return token
+
+    def request(self, method: str, resource: str, **kwargs) -> Response:
+        """Does an http request, if token is expired, then it will refresh"""
+        print(resource)
```
Comment (balloob) [inline ring_doorbell/auth.py]: stale print.

File: ring_doorbell/__init__.py (hunk 2 above)
Comment (balloob) [reply ring_doorbell/__init__.py]: In Home Assistant, the idea will be to no longer store username/password. It's only used on initial set up, after that we will just store the refresh/access tokens.

**Comment by JZ-SmartThings** [comment]
sorry for the lame question but what is the easiest way to manually fix this? Do I just grab the ring_doorbell folder contents and throw it under: **~/.homeassistant/deps/lib/python3.7/site-packages/ring_doorbell**

I also see __pycache__ in that folder? Should I wipe that folder or its contents then restart HA?

**Comment by steve-gombos** [comment]
Since it looks like the manifest was merged into the HA repo, I think now the easiest way is to just take a copy of the ring component from https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/ring and put it in your custom_components folder.

**Comment by balloob** [comment]
Update to 0.103.5

**Comment by steve-gombos** [comment]
Even easier. 👍

**Comment by Tosko4** [comment]
@balloob thanks for the info. Hopefully I can upgrade to 0.103.5 soon!

**Comment by JZ-SmartThings** [comment]
Worked like a charm... thanks @steve-gombos and @balloob!!!

File: ring_doorbell/__init__.py (hunk 2 above)
Comment (denniswham) [reply ring_doorbell/__init__.py]: Is this for a future version, or already implemented in the 103.5 version? I.e. do I still need to store the Ring password in HA, or is that not needed anymore?

Comment (tchellomello) [reply ring_doorbell/__init__.py]: It is still required to put the username and password into your HA configuration

**Pull request merged**